src/transformers/generation_beam_constraints.py

```diff
@@ -202,6 +211,170 @@ def copy(self, stateful=False):
         return new_constraint
 
 
+class DisjunctiveTrie:
+    def __init__(self, nested_token_ids):
+        r"""
+        A helper class that builds a trie with the words represented in `nested_token_ids`.
+
+        For example, if `nested_token_ids==[[1,2,3], [1,2,4], [1,5,6]]`, then the trie is: 1 -> 2 -> 3
+         \ -> 4
+          -> 5 -> 6
+        """
+        self.max_height = max([len(one) for one in nested_token_ids])
+
+        root = dict()
+        for token_ids in nested_token_ids:
+            level = root
+            for token_id in token_ids:
+                if token_id not in level:
+                    level[token_id] = dict()
+                level = level[token_id]
+
+            # indicate end of node
+            level[-1] = dict()
+
+        self.trie = root
+
+    def next_tokens(self, current_seq):
+        """
+        The next possible tokens that will progress the trie, given the current sequence of tokens in `current_seq`.
+
+        -1 in next_tokens refers to end of some constraint
+        """
+        start = self.trie
+
+        for current_token in current_seq:
+            current_token = current_token if isinstance(current_token, int) else current_token.tolist()
+            start = start[current_token]
+
+        next_tokens = list(start.keys())
+
+        return next_tokens
+
+    def reached_leaf(self, current_seq):
+        next_tokens = self.next_tokens(current_seq)
+
+        return -1 in next_tokens
+
+
+class DisjunctiveConstraint(Constraint):
+    r"""
+    A special [`Constraint`] that is fulfilled by fulfilling just one of several constraints.
+
+    Args:
+        constraints (`List[Constraint]`):
+            The list of [`Constraint`] objects, fulfillment of just one of which fulfills this
+            [`DisjunctiveConstraint`].
+    """
+
+    def __init__(self, nested_token_ids: List[Constraint]):
+        super(Constraint, self).__init__()
+
+        self.token_ids = nested_token_ids
+        self._check_valid(self.token_ids)
+
+        self.trie = DisjunctiveTrie([ids.tolist() for ids in self.token_ids])
+
+        self.seqlen = self.trie.max_height
+        self.current_seq = []
+        self.completed = False
+
+    def _check_valid(self, nested_token_ids):
+        """
+        Checks whether nested_token_ids is valid.
+        1. Checks Type
+        2. Checks for illegal subsets
+            We're checking if there is a `seq_k` in `nested_token_ids` such that `seq_k` is 100% a subset found in some
+            seq in `nested_token_ids`. We can't allow this because it leads to a perverse interpretation of
+            "completeness". Luckily, we only need to check the beginning of the sequence since it's a trie
+            representation.
+        """
+        if not isinstance(nested_token_ids, list):
+            nested_token_ids = nested_token_ids.tolist()
+            self.token_ids = nested_token_ids
```

Comment:
I don't think a function called `check_xxx` should ever have side effects, like changing class attributes.

If you want to do sanitation in that way, using a function `clean_xxx` that returns the cleaned objects and use
```python
def __init__(...):
     self.token_ids = self._clean_token_ids(token_ids)
```

is IMO much cleaner.

Reply:
Having read the full PR, I think most of this should be handled before this `__init__` (for the format of the argument) and this __init__ should only check for the "no subset of another token".